ui/components/component-library/modal-content/modal-content.tsx

```diff
@@ -1,37 +1,106 @@
-import React from 'react';
+import React, { forwardRef, useRef, useEffect } from 'react';
 import classnames from 'classnames';
 
 import {
   BackgroundColor,
   BorderRadius,
   BLOCK_SIZES,
+  DISPLAY,
+  JustifyContent,
+  AlignItems,
 } from '../../../helpers/constants/design-system';
 
 import Box from '../../ui/box/box';
 
+import { ModalFocus, useModalContext } from '..';
+
 import { ModalContentProps, ModalContentSize } from './modal-content.types';
 
-export const ModalContent = ({
-  className = '',
-  children,
-  size = ModalContentSize.Sm,
-  width,
-  modalContentRef, // Would have preferred to forwardRef but it's not trivial in TypeScript. Will update once we have an established pattern
-  ...props
-}: ModalContentProps) => (
-  <Box
-    className={classnames(
-      'mm-modal-content',
-      { [`mm-modal-content--size-${size}`]: !width },
-      className,
-    )}
-    backgroundColor={BackgroundColor.backgroundDefault}
-    borderRadius={BorderRadius.LG}
-    width={width || BLOCK_SIZES.FULL}
-    padding={4}
-    ref={modalContentRef}
-    {...props}
-  >
-    {children}
-  </Box>
+export const ModalContent = forwardRef(
+  (
+    {
+      className = '',
+      children,
+      size = ModalContentSize.Sm,
+      modalDialogProps,
+      ...props
+    }: ModalContentProps,
+    ref: React.Ref<HTMLElement>,
+  ) => {
+    const {
+      onClose,
+      isClosedOnEscapeKey,
+      isClosedOnOutsideClick,
+      initialFocusRef,
+      finalFocusRef,
+      restoreFocus,
+      autoFocus,
+    } = useModalContext();
+    const modalDialogRef = useRef<HTMLElement>(null);
+    const handleEscKey = (event: KeyboardEvent) => {
+      if (isClosedOnEscapeKey && event.key === 'Escape') {
+        onClose();
+      }
+    };
+
+    const handleClickOutside = (event: MouseEvent) => {
+      if (
+        isClosedOnOutsideClick &&
+        modalDialogRef?.current &&
+        !modalDialogRef.current.contains(event.target as Node)
+      ) {
+        onClose();
+      }
+    };
+
+    useEffect(() => {
```

Comment:
Hmm I don't think we would want to put the useEffect inside a conditional statement? useEffect is meant to handle side effects on every render or in response to specific dependencies. By putting it in a conditional it could lead to inconsistent behaviour depending on the condition and how it changes over renders. I could put the conditional inside of the useEffect and add `isClosedOnEscapeKey` and `isClosedOnOutsideClick` to the dependency array.  e.g.

```
  if (isClosedOnEscapeKey) {
    document.addEventListener('keydown', handleEscKey);
  }

  if (isClosedOnOutsideClick) {
    document.addEventListener('mousedown', handleClickOutside);
  }

  return () => {
    if (isClosedOnEscapeKey) {
      document.removeEventListener('keydown', handleEscKey);
    }

    if (isClosedOnOutsideClick) {
      document.removeEventListener('mousedown', handleClickOutside);
    }
  };
```

But I think the performance improvements would be pretty negligible? These event listeners would only be added and removed every time a Modal is open. But it would be good to monitor